File: tests/test_zero_waste_monkey.py
# -*- coding: utf-8 -*-
"""
test_zero_waste_monkey.py — ZeroWasteCache Adversarial / Monkey Tests
======================================================================

Targets: Core/zero_waste_cache.py (ZeroWasteCache, CacheFingerprint, Tier1/Tier2 entries)
Tests memory bounds, TTL boundaries, concurrent access, adversarial inputs.

Run:
    pytest tests/test_zero_waste_monkey.py -v --tb=short -x
"""

import random
import string
import sys
import threading
import time
from pathlib import Path
from typing import Dict, List
from unittest.mock import MagicMock

import pytest

ROOT = Path(__file__).resolve().parent.parent
sys.path.insert(0, str(ROOT))

_CHAOS_STRINGS: list[str] = [
    "",
    "   ",
    "\x00\x01\x02",
    "A" * 100_000,
    "🔥" * 5_000,
    "<script>alert('xss')</script>",
    "'; DROP TABLE users; --",
    "Hello 你好 مرحبا こんにちは",
    "\u202e\u200b\u200c\u200d\ufeff",
    "NaN",
    "null",
    "None",
    "what is the latest news",
    "today's weather",
    "current stock price",
]


def _mock_model():
    model = MagicMock()
    model.encode.return_value = [random.random() for _ in range(384)]
    return model


# ═════════════════════════════════════════════════════════════════════════════
#  CacheFingerprint Tests
# ═════════════════════════════════════════════════════════════════════════════


@pytest.mark.monkey
class TestCacheFingerprintMonkey:
    """Verify CacheFingerprint creation and from_chunks factory."""

    def test_from_chunks_empty(self):
        from Core.zero_waste_cache import CacheFingerprint

        fp = CacheFingerprint.from_chunks([], collection_version=0)
        assert isinstance(fp.combined_hash, str)
        assert len(fp.chunk_hashes) == 0

    def test_from_chunks_normal(self):
        from Core.zero_waste_cache import CacheFingerprint

        chunks = [
            {"text": "chunk 1", "source_url": "http://example.com/1"},
            {"text": "chunk 2", "source_url": "http://example.com/2"},
        ]
        fp = CacheFingerprint.from_chunks(chunks, collection_version=1)
        assert len(fp.chunk_hashes) == 2
        assert fp.collection_version == 1

    def test_from_chunks_no_text_key(self):
        """Chunks without 'text' key — should handle gracefully."""
        from Core.zero_waste_cache import CacheFingerprint

        chunks = [{"id": 1}, {"id": 2}]
        try:
            fp = CacheFingerprint.from_chunks(chunks, collection_version=0)
            assert isinstance(fp, CacheFingerprint)
        except (KeyError, TypeError):
            pass  # raising on missing 'text' is acceptable


# ═════════════════════════════════════════════════════════════════════════════
#  CacheValidationStrategy Tests
# ═════════════════════════════════════════════════════════════════════════════


@pytest.mark.monkey
class TestCacheValidationStrategyMonkey:
    """Verify strategy enum and classify_strategy."""

    def test_all_strategies_exist(self):
        from Core.zero_waste_cache import CacheValidationStrategy

        expected = {"FINGERPRINT", "TEMPORAL", "VERSION", "STATIC"}
        actual = {s.name for s in CacheValidationStrategy}
        assert expected.issubset(actual)

    def test_classify_temporal_queries(self):
        cache = self._make_cache()
        temporal_queries = ["what is today's news", "latest updates", "current weather"]
        for q in temporal_queries:
            strategy = cache.classify_strategy(q)
            # Should recognize temporal keywords
            from Core.zero_waste_cache import CacheValidationStrategy

            assert isinstance(strategy, CacheValidationStrategy)

    def test_classify_static_queries(self):
        cache = self._make_cache()
        static_queries = ["what is photosynthesis", "explain quantum mechanics"]
        for q in static_queries:
            strategy = cache.classify_strategy(q)
            from Core.zero_waste_cache import CacheValidationStrategy

            assert isinstance(strategy, CacheValidationStrategy)

    def test_classify_chaos_strings(self):
        cache = self._make_cache()
        for s in _CHAOS_STRINGS:
            try:
                strategy = cache.classify_strategy(s)
                from Core.zero_waste_cache import CacheValidationStrategy

                assert isinstance(strategy, CacheValidationStrategy)
            except (ValueError, TypeError):
                pass  # rejecting garbage is fine

    def _make_cache(self):
        from Core.zero_waste_cache import ZeroWasteCache

        return ZeroWasteCache(model=_mock_model(), max_entries=100)


# ═════════════════════════════════════════════════════════════════════════════
#  ZeroWasteCache Core Tests
# ═════════════════════════════════════════════════════════════════════════════


@pytest.mark.monkey
class TestZeroWasteCacheMonkey:
    """Adversarial tests for the 2-tier cache."""

    def _make_cache(self, max_entries=100, ttl1=3600, ttl2=7200):
        from Core.zero_waste_cache import ZeroWasteCache

        return ZeroWasteCache(
            model=_mock_model(),
            max_entries=max_entries,
            tier1_ttl=ttl1,
            tier2_ttl=ttl2,
        )

    def test_set_get_roundtrip(self):
        cache = self._make_cache()
        chunks = [{"text": "chunk data", "score": 0.9}]
        cache.set_answer("test query", results=[{"text": "answer"}], source_chunks=chunks)
        result = cache.get_answer("test query")
        # May or may not hit due to threshold, but must not crash

    def test_chaos_strings_no_crash(self):
        cache = self._make_cache()
        for s in _CHAOS_STRINGS:
            try:
                cache.set_answer(s, results=[{"text": "ans"}], source_chunks=[{"text": "chunk"}])
                cache.get_answer(s)
                cache.set_context(s, chunks=[{"text": "ctx"}])
                cache.get_context(s)
            except (TypeError, ValueError):
                pass  # rejecting bad input is fine

    def test_1000_items_memory_bounded(self):
        """Insert 1000 items into a 100-entry cache — must evict, not OOM."""
        cache = self._make_cache(max_entries=100)
        for i in range(1000):
            cache.set_answer(
                f"query_{i}",
                results=[{"text": f"answer_{i}"}],
                source_chunks=[{"text": f"chunk_{i}"}],
            )
        # If internal size tracking exists, check it
        # Otherwise just verify no crash

    def test_invalidate_urls(self):
        cache = self._make_cache()
        cache.set_answer(
            "test",
            results=[{"text": "ans"}],
            source_chunks=[{"text": "chunk", "source_url": "http://example.com"}],
        )
        count = cache.invalidate_urls({"http://example.com"})
        assert isinstance(count, int)
        assert count >= 0

    def test_invalidate_empty_urls(self):
        cache = self._make_cache()
        count = cache.invalidate_urls(set())
        assert count == 0

    def test_bump_version(self):
        cache = self._make_cache()
        cache.bump_version()
        cache.bump_version()  # double bump must not crash

    def test_is_temporal_query(self):
        cache = self._make_cache()
        assert cache.is_temporal_query("what is today's news") or True  # result depends on impl
        assert isinstance(cache.is_temporal_query("explain gravity"), bool)

    def test_zero_ttl(self):
        """TTL=0 means immediate expiry — get should return None."""
        cache = self._make_cache(ttl1=0, ttl2=0)
        cache.set_answer("q", results=[{"text": "a"}], source_chunks=[{"text": "c"}])
        # Immediate get — may or may not be expired depending on implementation
        result = cache.get_answer("q")
        assert result is None or isinstance(result, (list, tuple, dict))

    def test_nan_embedding_no_crash(self):
        """Model returns NaN embeddings — cache must handle gracefully."""
        from Core.zero_waste_cache import ZeroWasteCache

        model = MagicMock()
        model.encode.return_value = [float("nan")] * 384
        cache = ZeroWasteCache(model=model, max_entries=50)
        try:
            cache.set_answer("test", results=[{"text": "a"}], source_chunks=[{"text": "c"}])
            cache.get_answer("test")
        except (ValueError, RuntimeError):
            pass  # raising on NaN is acceptable


# ═════════════════════════════════════════════════════════════════════════════
#  Concurrent Access Tests
# ═════════════════════════════════════════════════════════════════════════════


@pytest.mark.monkey
class TestZeroWasteConcurrencyMonkey:
    """Thread safety for ZeroWasteCache get/set operations."""

    def test_concurrent_get_set(self):
        from Core.zero_waste_cache import ZeroWasteCache

        cache = ZeroWasteCache(model=_mock_model(), max_entries=200)
        errors = []

        def writer(tid):
            try:
                for i in range(100):
                    cache.set_answer(
                        f"t{tid}_q{i}",
                        results=[{"text": f"ans_{tid}_{i}"}],
                        source_chunks=[{"text": f"chunk_{tid}_{i}"}],
                    )
            except Exception as e:
                errors.append(e)

        def reader(tid):
            try:
                for i in range(100):
                    cache.get_answer(f"t{tid}_q{random.randint(0, 99)}")
            except Exception as e:
                errors.append(e)

        threads = []
        for t in range(5):
            threads.append(threading.Thread(target=writer, args=(t,)))
            threads.append(threading.Thread(target=reader, args=(t,)))

        for t in threads:
            t.start()
        for t in threads:
            t.join(timeout=30)

        assert not errors, f"Concurrency errors: {errors}"

    def test_concurrent_invalidate(self):
        from Core.zero_waste_cache import ZeroWasteCache

        cache = ZeroWasteCache(model=_mock_model(), max_entries=200)
        errors = []

        # Fill cache
        for i in range(50):
            cache.set_answer(
                f"q{i}",
                results=[{"text": f"a{i}"}],
                source_chunks=[{"text": f"c{i}", "source_url": f"http://example.com/{i}"}],
            )

        def invalidator(tid):
            try:
                for i in range(50):
                    cache.invalidate_urls({f"http://example.com/{random.randint(0, 49)}"})
            except Exception as e:
                errors.append(e)

        threads = [threading.Thread(target=invalidator, args=(t,)) for t in range(5)]
        for t in threads:
            t.start()
        for t in threads:
            t.join(timeout=15)

        assert not errors, f"Invalidation errors: {errors}"

File: rust_output/src/test_zero_waste_monkey.rs
/// test_zero_waste_monkey::py — ZeroWasteCache Adversarial / Monkey Tests
/// ======================================================================
/// 
/// Targets: Core/zero_waste_cache::py (ZeroWasteCache, CacheFingerprint, Tier1/Tier2 entries)
/// Tests memory bounds, TTL boundaries, concurrent access, adversarial inputs.
/// 
/// Run:
/// pytest tests/test_zero_waste_monkey::py -v --tb=short -x

use anyhow::{Result, Context};
use std::collections::HashMap;
use std::collections::HashSet;

pub const ROOT: &str = "Path(file!()).resolve().parent.parent";

pub static _CHAOS_STRINGS: std::sync::LazyLock<Vec<String>> = std::sync::LazyLock::new(|| Vec::new());

/// Verify CacheFingerprint creation and from_chunks factory.
#[derive(Debug, Clone)]
pub struct TestCacheFingerprintMonkey {
}

impl TestCacheFingerprintMonkey {
    pub fn test_from_chunks_empty(&self) -> () {
        // TODO: from Core.zero_waste_cache import CacheFingerprint
        let mut fp = CacheFingerprint.from_chunks(vec![], /* collection_version= */ 0);
        assert!(/* /* isinstance(fp.combined_hash, str) */ */ true);
        assert!(fp.chunk_hashes.len() == 0);
    }
    pub fn test_from_chunks_normal(&self) -> () {
        // TODO: from Core.zero_waste_cache import CacheFingerprint
        let mut chunks = vec![HashMap::from([("text".to_string(), "chunk 1".to_string()), ("source_url".to_string(), "http://example.com/1".to_string())]), HashMap::from([("text".to_string(), "chunk 2".to_string()), ("source_url".to_string(), "http://example.com/2".to_string())])];
        let mut fp = CacheFingerprint.from_chunks(chunks, /* collection_version= */ 1);
        assert!(fp.chunk_hashes.len() == 2);
        assert!(fp.collection_version == 1);
    }
    /// Chunks without 'text' key — should handle gracefully.
    pub fn test_from_chunks_no_text_key(&self) -> Result<()> {
        // Chunks without 'text' key — should handle gracefully.
        // TODO: from Core.zero_waste_cache import CacheFingerprint
        let mut chunks = vec![HashMap::from([("id".to_string(), 1)]), HashMap::from([("id".to_string(), 2)])];
        // try:
        {
            let mut fp = CacheFingerprint.from_chunks(chunks, /* collection_version= */ 0);
            assert!(/* /* isinstance(fp, CacheFingerprint) */ */ true);
        }
        // except (KeyError, TypeError) as _e:
    }
}

/// Verify strategy enum and classify_strategy.
#[derive(Debug, Clone)]
pub struct TestCacheValidationStrategyMonkey {
}

impl TestCacheValidationStrategyMonkey {
    pub fn test_all_strategies_exist(&self) -> () {
        // TODO: from Core.zero_waste_cache import CacheValidationStrategy
        let mut expected = HashSet::from(["FINGERPRINT".to_string(), "TEMPORAL".to_string(), "VERSION".to_string(), "STATIC".to_string()]);
        let mut actual = CacheValidationStrategy.iter().map(|s| s.name).collect::<HashSet<_>>();
        assert!(expected.issubset(actual));
    }
    pub fn test_classify_temporal_queries(&mut self) -> () {
        let mut cache = self._make_cache();
        let mut temporal_queries = vec!["what is today's news".to_string(), "latest updates".to_string(), "current weather".to_string()];
        for q in temporal_queries.iter() {
            let mut strategy = cache::classify_strategy(q);
            // TODO: from Core.zero_waste_cache import CacheValidationStrategy
            assert!(/* /* isinstance(strategy, CacheValidationStrategy) */ */ true);
        }
    }
    pub fn test_classify_static_queries(&mut self) -> () {
        let mut cache = self._make_cache();
        let mut static_queries = vec!["what is photosynthesis".to_string(), "explain quantum mechanics".to_string()];
        for q in static_queries.iter() {
            let mut strategy = cache::classify_strategy(q);
            // TODO: from Core.zero_waste_cache import CacheValidationStrategy
            assert!(/* /* isinstance(strategy, CacheValidationStrategy) */ */ true);
        }
    }
    pub fn test_classify_chaos_strings(&mut self) -> Result<()> {
        let mut cache = self._make_cache();
        for s in _CHAOS_STRINGS.iter() {
            // try:
            {
                let mut strategy = cache::classify_strategy(s);
                // TODO: from Core.zero_waste_cache import CacheValidationStrategy
                assert!(/* /* isinstance(strategy, CacheValidationStrategy) */ */ true);
            }
            // except (ValueError, TypeError) as _e:
        }
    }
    pub fn _make_cache(&self) -> () {
        // TODO: from Core.zero_waste_cache import ZeroWasteCache
        ZeroWasteCache(/* model= */ _mock_model(), /* max_entries= */ 100)
    }
}

/// Adversarial tests for the 2-tier cache.
#[derive(Debug, Clone)]
pub struct TestZeroWasteCacheMonkey {
}

impl TestZeroWasteCacheMonkey {
    pub fn _make_cache(&self, max_entries: String, ttl1: String, ttl2: String) -> () {
        // TODO: from Core.zero_waste_cache import ZeroWasteCache
        ZeroWasteCache(/* model= */ _mock_model(), /* max_entries= */ max_entries, /* tier1_ttl= */ ttl1, /* tier2_ttl= */ ttl2)
    }
    pub fn test_set_get_roundtrip(&mut self) -> () {
        let mut cache = self._make_cache();
        let mut chunks = vec![HashMap::from([("text".to_string(), "chunk data".to_string()), ("score".to_string(), 0.9_f64)])];
        cache::set_answer("test query".to_string(), /* results= */ vec![HashMap::from([("text".to_string(), "answer".to_string())])], /* source_chunks= */ chunks);
        let mut result = cache::get_answer("test query".to_string());
    }
    pub fn test_chaos_strings_no_crash(&mut self) -> Result<()> {
        let mut cache = self._make_cache();
        for s in _CHAOS_STRINGS.iter() {
            // try:
            {
                cache::set_answer(s, /* results= */ vec![HashMap::from([("text".to_string(), "ans".to_string())])], /* source_chunks= */ vec![HashMap::from([("text".to_string(), "chunk".to_string())])]);
                cache::get_answer(s);
                cache::set_context(s, /* chunks= */ vec![HashMap::from([("text".to_string(), "ctx".to_string())])]);
                cache::get_context(s);
            }
            // except (TypeError, ValueError) as _e:
        }
    }
    /// Insert 1000 items into a 100-entry cache — must evict, not OOM.
    pub fn test_1000_items_memory_bounded(&mut self) -> () {
        // Insert 1000 items into a 100-entry cache — must evict, not OOM.
        let mut cache = self._make_cache(/* max_entries= */ 100);
        for i in 0..1000.iter() {
            cache::set_answer(format!("query_{}", i), /* results= */ vec![HashMap::from([("text".to_string(), format!("answer_{}", i))])], /* source_chunks= */ vec![HashMap::from([("text".to_string(), format!("chunk_{}", i))])]);
        }
    }
    pub fn test_invalidate_urls(&mut self) -> () {
        let mut cache = self._make_cache();
        cache::set_answer("test".to_string(), /* results= */ vec![HashMap::from([("text".to_string(), "ans".to_string())])], /* source_chunks= */ vec![HashMap::from([("text".to_string(), "chunk".to_string()), ("source_url".to_string(), "http://example.com".to_string())])]);
        let mut count = cache::invalidate_urls(HashSet::from(["http://example.com".to_string()]));
        assert!(/* /* isinstance(count, int) */ */ true);
        assert!(count >= 0);
    }
    pub fn test_invalidate_empty_urls(&mut self) -> () {
        let mut cache = self._make_cache();
        let mut count = cache::invalidate_urls(HashSet::new());
        assert!(count == 0);
    }
    pub fn test_bump_version(&mut self) -> () {
        let mut cache = self._make_cache();
        cache::bump_version();
        cache::bump_version();
    }
    pub fn test_is_temporal_query(&mut self) -> () {
        let mut cache = self._make_cache();
        assert!((cache::is_temporal_query("what is today's news".to_string()) || true));
        assert!(/* /* isinstance(cache::is_temporal_query("explain gravity".to_string()), bool) */ */ true);
    }
    /// TTL=0 means immediate expiry — get should return None.
    pub fn test_zero_ttl(&mut self) -> () {
        // TTL=0 means immediate expiry — get should return None.
        let mut cache = self._make_cache(/* ttl1= */ 0, /* ttl2= */ 0);
        cache::set_answer("q".to_string(), /* results= */ vec![HashMap::from([("text".to_string(), "a".to_string())])], /* source_chunks= */ vec![HashMap::from([("text".to_string(), "c".to_string())])]);
        let mut result = cache::get_answer("q".to_string());
        assert!((result.is_none() || /* /* isinstance(result, (list, tuple, dict) */) */ true));
    }
    /// Model returns NaN embeddings — cache must handle gracefully.
    pub fn test_nan_embedding_no_crash(&self) -> Result<()> {
        // Model returns NaN embeddings — cache must handle gracefully.
        // TODO: from Core.zero_waste_cache import ZeroWasteCache
        let mut model = MagicMock();
        model.encode.return_value = (vec!["nan".to_string().to_string().parse::<f64>().unwrap_or(0.0)] * 384);
        let mut cache = ZeroWasteCache(/* model= */ model, /* max_entries= */ 50);
        // try:
        {
            cache::set_answer("test".to_string(), /* results= */ vec![HashMap::from([("text".to_string(), "a".to_string())])], /* source_chunks= */ vec![HashMap::from([("text".to_string(), "c".to_string())])]);
            cache::get_answer("test".to_string());
        }
        // except (ValueError, RuntimeError) as _e:
    }
}

/// Thread safety for ZeroWasteCache get/set operations.
#[derive(Debug, Clone)]
pub struct TestZeroWasteConcurrencyMonkey {
}

impl TestZeroWasteConcurrencyMonkey {
    pub fn test_concurrent_get_set(&self) -> Result<()> {
        // TODO: from Core.zero_waste_cache import ZeroWasteCache
        let mut cache = ZeroWasteCache(/* model= */ _mock_model(), /* max_entries= */ 200);
        let mut errors = vec![];
        let writer = |tid| {
            // try:
            {
                for i in 0..100.iter() {
                    cache::set_answer(format!("t{}_q{}", tid, i), /* results= */ vec![HashMap::from([("text".to_string(), format!("ans_{}_{}", tid, i))])], /* source_chunks= */ vec![HashMap::from([("text".to_string(), format!("chunk_{}_{}", tid, i))])]);
                }
            }
            // except Exception as e:
        };
        let reader = |tid| {
            // try:
            {
                for i in 0..100.iter() {
                    cache::get_answer(format!("t{}_q{}", tid, random.randint(0, 99)));
                }
            }
            // except Exception as e:
        };
        let mut threads = vec![];
        for t in 0..5.iter() {
            threads.push(std::thread::spawn(|| {}));
            threads.push(std::thread::spawn(|| {}));
        }
        for t in threads.iter() {
            t.start();
        }
        for t in threads.iter() {
            /* timeout= */ 30.join(&t);
        }
        Ok(assert!(!errors, "Concurrency errors: {}", errors))
    }
    pub fn test_concurrent_invalidate(&self) -> Result<()> {
        // TODO: from Core.zero_waste_cache import ZeroWasteCache
        let mut cache = ZeroWasteCache(/* model= */ _mock_model(), /* max_entries= */ 200);
        let mut errors = vec![];
        for i in 0..50.iter() {
            cache::set_answer(format!("q{}", i), /* results= */ vec![HashMap::from([("text".to_string(), format!("a{}", i))])], /* source_chunks= */ vec![HashMap::from([("text".to_string(), format!("c{}", i)), ("source_url".to_string(), format!("http://example.com/{}", i))])]);
        }
        let invalidator = |tid| {
            // try:
            {
                for i in 0..50.iter() {
                    cache::invalidate_urls(HashSet::from([format!("http://example.com/{}", random.randint(0, 49))]));
                }
            }
            // except Exception as e:
        };
        let mut threads = 0..5.iter().map(|t| std::thread::spawn(|| {})).collect::<Vec<_>>();
        for t in threads.iter() {
            t.start();
        }
        for t in threads.iter() {
            /* timeout= */ 15.join(&t);
        }
        Ok(assert!(!errors, "Invalidation errors: {}", errors))
    }
}

pub fn _mock_model() -> () {
    let mut model = MagicMock();
    model.encode.return_value = 0..384.iter().map(|_| random.random()).collect::<Vec<_>>();
    model
}
